stumpy/stumpi.py

```diff
@@ -287,54 +311,62 @@ def _update(self, t):
 
         core.apply_exclusion_zone(D, D.shape[0] - 1, self._excl_zone, np.inf)
 
-        update_idx = np.argwhere(D[:l] < self._P[:l]).flatten()
-        self._I[update_idx] = l
-        self._P[update_idx] = D[update_idx]
+        update_idx = np.argwhere(D[:l] < self._P[:l, -1]).flatten()
+        for i in update_idx:
+            pos = np.searchsorted(self._P[i], D[i], side="right")
+            core._shift_insert_at_index(self._P[i], pos, D[i])
+            core._shift_insert_at_index(self._I[i], pos, l)
 
-        I_last = np.argmin(D)
-        if np.isinf(D[I_last]):
-            I_new = np.append(self._I, -1)
-            P_new = np.append(self._P, np.inf)
-        else:
-            I_new = np.append(self._I, I_last)
-            P_new = np.append(self._P, D[I_last])
-        left_I_new = np.append(self._left_I, I_last)
-        left_P_new = np.append(self._left_P, D[I_last])
+        # Calculating top-k and left matrix profile for new subsequence whose
+        # distance profie is D
+
+        # O(Nlog(K)) time complexity for obtaining top-k
+        P_new = np.full(self._k, np.inf, dtype=np.float64)
+        I_new = np.full(self._k, -1, dtype=np.int64)
+
+        for i, d in enumerate(D):
+            if d < P_new[-1]:  # maximum value in sorted array P_new
+                pos = np.searchsorted(P_new, d, side="right")
+                core._shift_insert_at_index(P_new, pos, d)
+                core._shift_insert_at_index(I_new, pos, i)
+
+        left_I_new = I_new[0]
+        left_P_new = P_new[0]
 
         self._T = T_new
-        self._P = P_new
-        self._I = I_new
-        self._left_I = left_I_new
-        self._left_P = left_P_new
+        self._P = np.append(self._P, P_new.reshape(1, -1), axis=0)
+        self._I = np.append(self._I, I_new.reshape(1, -1), axis=0)
+        self._left_P = np.append(self._left_P, left_P_new)
+        self._left_I = np.append(self._left_I, left_I_new)
         self._QT = QT_new
         self._M_T = M_T_new
         self._Σ_T = Σ_T_new
 
     @property
     def P_(self):
         """
-        Get the matrix profile
+        Get the (top-k) matrix profile
         """
         return self._P.astype(np.float64)
```

Comment:
Hmmm, this is a breaking change! I am concerned that it will cause problems in other people's code. At the same time, the inconsistency of returning a 1D array when `k=1` but a 2D array when `k>1` would really bother me! 

Another alternative (not saying it is better) is to leave `P_` to always return the top-1 1D matrix profile and we add a new attribute that always returns the top-k 2D matrix profile (regardless of `k`)

Reply:
> Another alternative (not saying it is better) is to leave `P_` to always return the top-1 1D matrix profile and we add a new attribute that always returns the top-k 2D matrix profile (regardless of `k`)

I think this is a good idea. From my perspective, the only downside is that the `P_` is not `top-k` anymore, which is in contrast to what we have in mind about `P`. However, this solution will allow users to continue working with STUMPY with no problem.

---

Note that `stumpy/scrump.py` suffers from the same issue. However, since you said a few users may use `scrump`, I assume you are okay with this problem in `scrump`.

Reply:
Yeah, I am less concerned about `scrump`. I had a feeling that `stumpi` would break and hence why I asked about the tutorials :)

Reply:
There is one more solution. In the case of a breaking change, we increase the "major" version from `1.11.x` to `2.0.0` which implies that we had to introduce breaking changes. Then, we make EVERYTHING 2D and nothing is 1D. 

Is there anything else that is still currently 1D?

Reply:
> Then, we make EVERYTHING 2D and nothing is 1D.
> 
> Is there anything else that is still currently 1D?

I think `scrump` and `stumpi` are the only ones that give 2D array when `k=1`.

By EVERYTHING, did you mean all modules? So, in that case, please recall that the output of `stumpy/stump.py` is 2D even for `k=1`. However, if we try to get `P`, then: `P=mp[:, 0]` would be 1D. Is that okay?

Reply:
So, I think we need to figure out this first and then, I need to run those tutorials again to make sure everything is fine.

Reply:
> By EVERYTHING, did you mean all modules? So, in that case, please recall that the output of stumpy/stump.py is 2D even for k=1.

Yes, all modules. 

>  However, if we try to get P, then: P=mp[:, 0] would be 1D. Is that okay?

Yes, that's fine.

> So, I think we need to figure out this first and then, I need to run those tutorials again to make sure everything is fine.

I think the best option is to always return 2D even when `k=1` for all modules (except for `stimp`). Then we bump our major version to `2.0.0` in the next release. I think it is okay due to all of the `top-k` work that will make things "different". Though, we shouldn't make a habit of bumping the major version.

Reply:
> I can checkout other modules as well to see if they return P and/or I, and make sure they are 2D even for k=1.

That would be great! I'm thinking whether there are any occasions when me might change left/right matrix profiles to be top-k as well. If so, then we may want to also allows those arrays to be 2D by default. Just trying to anticipate the future

Reply:
> That would be great! I'm thinking whether there are any occasions when me might change left/right matrix profiles to be top-k as well. 

That is possible! (e.g. someone may try the concept of VALMOD on left/right neighbors. So, returning top-k for left/right might be useful in the future).

>  If so, then we may want to also allows those arrays to be 2D by default

This means that we are going to increase the number of  changed files here. Because I need to consider normalized / non-normalized functions AND left/right matrix profile/index. Also, I need to probably change their test functions as well. Also, we need to change the modules that use `P` / `I` as inputs. (e.g. `chains.py`) 

Even if I do not enhance the non-normalized functions to support top-k for now, I need to revise them to return 2D arrays...
```
P = np.full((l, 1), np.inf, dtype=np.float64)
```

So....this PR can explode! Should we do it all here?

Reply:
So, I propose that we split it up into the following PRs:

1. top-k for normalized methods (this PR)
2. top-k for non-normalized methods
3. 2D arrays (output) returned everywhere

What do you think? 

> Also, we need to change the modules that use P / I as inputs. (e.g. chains.py)

I don't know, I think we can just leave the inputs as 1D. Are there other functions beyond those in `chains.py`?

Reply:
> I propose that we split it up into the following PRs:

I think this would be a good idea. This way we can better focus on some related tasks and reduce the chance of missing something!

> I don't know, I think we can just leave the inputs as 1D

> Are there other functions beyond those in chains.py?

I am not sure. I just wanted to give an example. But, I think this will be more clear when I dig into it. So, if you do not mind, I am going to wrap up the first item `1. top-k for normalized methods (this PR)`. Then, after finishing the second one, i.e. non-normalized methods, which I think it should be easy, I can focus on the third one. 

Are you okay with this?